src/core/ddsi/src/ddsi_cdrstream.c

```diff
@@ -895,10 +900,10 @@ static bool insn_key_ok_p (uint32_t insn)
 static uint32_t read_union_discriminant (dds_istream_t * __restrict is, uint32_t insn)
 {
   enum dds_stream_typecode type = DDS_OP_SUBTYPE (insn);
-  assert (type == DDS_OP_VAL_1BY || type == DDS_OP_VAL_2BY || type == DDS_OP_VAL_4BY || type == DDS_OP_VAL_ENU);
+  assert (is_primitive_type (type) || type == DDS_OP_VAL_ENU);
```

Comment:
The introduction of `is_primitive_type` is very useful, I think the code would further benefit from also adding `is_primitive_or_enum_type`.